### DISCUSSION TITLE
Native libraries?

### DISCUSSION THREAD

**Discussion by fafayqa** [beehive-lab/TornadoVM#140] [Q&A]
I am trying to load and test TornadoVM src in NetBeans 12.4 x64, I had to recreate whole src structure on my own as POM for some reason does not work for me (project is loaded, but no dependencies are solved/downloaded and a lot of other errors).
So I was "playing" with it whole day, after a head-ache situations I guess I made it to the point where all seems to be OK, nothing is missing (or it seems so in NetBeans = no errors of missing jars etc.).

But most of the examples ends up in errors , mostly with `Exception in thread "AWT-EventQueue-0" java.lang.NullPointerException`.

Those examples seems to be everytime such that are trying to use TaskSchedule().

For instance, example file/class BlackAndWhiteTransform.java: normally, when I just run the code as it is, it works great - it process the image. I think what it does is that it checks for itself if GPU can be used and if not switch to CPU. So I manually set it to do parallelComputation(g); instead of sequentialComputation(g); which it does on its own by default in my case. As soon as I did that and hit RUN, it ended up with lots of errors in Output window, like this one here:

```
run:
Exception in thread "AWT-EventQueue-0" java.lang.NullPointerException
	at java.lang.Class.forName0(Native Method)
	at java.lang.Class.forName(Class.java:264)
	at uk.ac.manchester.tornado.api.runtime.TornadoAPIProvider.loadScheduleRuntime(TornadoAPIProvider.java:62)
	at uk.ac.manchester.tornado.api.TaskSchedule.<init>(TaskSchedule.java:86)
	at uk.ac.manchester.tornado.examples.compute.BlackAndWhiteTransform$LoadImage.parallelComputation(BlackAndWhiteTransform.java:132)
	at uk.ac.manchester.tornado.examples.compute.BlackAndWhiteTransform$LoadImage.paint(BlackAndWhiteTransform.java:197)
	at javax.swing.JComponent.paintChildren(JComponent.java:903)
	at javax.swing.JComponent.paint(JComponent.java:1065)
	at javax.swing.JComponent.paintChildren(JComponent.java:889)
	at javax.swing.JComponent.paint(JComponent.java:1065)
	at javax.swing.JLayeredPane.paint(JLayeredPane.java:586)
	at javax.swing.JComponent.paintChildren(JComponent.java:889)
	at javax.swing.JComponent.paintToOffscreen(JComponent.java:5217)
	at javax.swing.BufferStrategyPaintManager.paint(BufferStrategyPaintManager.java:290)
	at javax.swing.RepaintManager.paint(RepaintManager.java:1272)
	at javax.swing.JComponent.paint(JComponent.java:1042)
	at java.awt.GraphicsCallback$PaintCallback.run(GraphicsCallback.java:39)
	at sun.awt.SunGraphicsCallback.runOneComponent(SunGraphicsCallback.java:79)
	at sun.awt.SunGraphicsCallback.runComponents(SunGraphicsCallback.java:116)
	at java.awt.Container.paint(Container.java:1978)
	at java.awt.Window.paint(Window.java:3906)
	at javax.swing.RepaintManager$4.run(RepaintManager.java:842)
	at javax.swing.RepaintManager$4.run(RepaintManager.java:814)
	at java.security.AccessController.doPrivileged(Native Method)
	at java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(ProtectionDomain.java:74)
	at javax.swing.RepaintManager.paintDirtyRegions(RepaintManager.java:814)
	at javax.swing.RepaintManager.paintDirtyRegions(RepaintManager.java:789)
	at javax.swing.RepaintManager.prePaintDirtyRegions(RepaintManager.java:738)
	at javax.swing.RepaintManager.access$1200(RepaintManager.java:64)
	at javax.swing.RepaintManager$ProcessingRunnable.run(RepaintManager.java:1732)
	at java.awt.event.InvocationEvent.dispatch(InvocationEvent.java:311)
	at java.awt.EventQueue.dispatchEventImpl(EventQueue.java:758)
	at java.awt.EventQueue.access$500(EventQueue.java:97)
	at java.awt.EventQueue$3.run(EventQueue.java:709)
	at java.awt.EventQueue$3.run(EventQueue.java:703)
	at java.security.AccessController.doPrivileged(Native Method)
	at java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(ProtectionDomain.java:74)
	at java.awt.EventQueue.dispatchEvent(EventQueue.java:728)
	at java.awt.EventDispatchThread.pumpOneEventForFilters(EventDispatchThread.java:205)
	at java.awt.EventDispatchThread.pumpEventsForFilter(EventDispatchThread.java:116)
	at java.awt.EventDispatchThread.pumpEventsForHierarchy(EventDispatchThread.java:105)
	at java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:101)
	at java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:93)
	at java.awt.EventDispatchThread.run(EventDispatchThread.java:82)
Exception in thread "AWT-EventQueue-0" java.lang.NullPointerException
	at java.lang.Class.forName0(Native Method)
	at java.lang.Class.forName(Class.java:264)
	at uk.ac.manchester.tornado.api.runtime.TornadoAPIProvider.loadScheduleRuntime(TornadoAPIProvider.java:62)
	at uk.ac.manchester.tornado.api.TaskSchedule.<init>(TaskSchedule.java:86)
	at uk.ac.manchester.tornado.examples.compute.BlackAndWhiteTransform$LoadImage.parallelComputation(BlackAndWhiteTransform.java:132)
	at uk.ac.manchester.tornado.examples.compute.BlackAndWhiteTransform$LoadImage.paint(BlackAndWhiteTransform.java:197)
	at javax.swing.JComponent.paintChildren(JComponent.java:903)
	at javax.swing.JComponent.paint(JComponent.java:1065)
	at javax.swing.JComponent.paintChildren(JComponent.java:889)
	at javax.swing.JComponent.paint(JComponent.java:1065)
	at javax.swing.JLayeredPane.paint(JLayeredPane.java:586)
	at javax.swing.JComponent.paintChildren(JComponent.java:889)
	at javax.swing.JComponent.paintToOffscreen(JComponent.java:5217)
	at javax.swing.BufferStrategyPaintManager.paint(BufferStrategyPaintManager.java:290)
	at javax.swing.RepaintManager.paint(RepaintManager.java:1272)
	at javax.swing.JComponent.paint(JComponent.java:1042)
	at java.awt.GraphicsCallback$PaintCallback.run(GraphicsCallback.java:39)
	at sun.awt.SunGraphicsCallback.runOneComponent(SunGraphicsCallback.java:79)
	at sun.awt.SunGraphicsCallback.runComponents(SunGraphicsCallback.java:116)
	at java.awt.Container.paint(Container.java:1978)
	at java.awt.Window.paint(Window.java:3906)
	at javax.swing.RepaintManager$4.run(RepaintManager.java:842)
	at javax.swing.RepaintManager$4.run(RepaintManager.java:814)
	at java.security.AccessController.doPrivileged(Native Method)
	at java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(ProtectionDomain.java:74)
	at javax.swing.RepaintManager.paintDirtyRegions(RepaintManager.java:814)
	at javax.swing.RepaintManager.paintDirtyRegions(RepaintManager.java:789)
	at javax.swing.RepaintManager.prePaintDirtyRegions(RepaintManager.java:738)
	at javax.swing.RepaintManager.access$1200(RepaintManager.java:64)
	at javax.swing.RepaintManager$ProcessingRunnable.run(RepaintManager.java:1732)
	at java.awt.event.InvocationEvent.dispatch(InvocationEvent.java:311)
	at java.awt.EventQueue.dispatchEventImpl(EventQueue.java:758)
	at java.awt.EventQueue.access$500(EventQueue.java:97)
	at java.awt.EventQueue$3.run(EventQueue.java:709)
	at java.awt.EventQueue$3.run(EventQueue.java:703)
	at java.security.AccessController.doPrivileged(Native Method)
	at java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(ProtectionDomain.java:74)
	at java.awt.EventQueue.dispatchEvent(EventQueue.java:728)
	at java.awt.EventDispatchThread.pumpOneEventForFilters(EventDispatchThread.java:205)
	at java.awt.EventDispatchThread.pumpEventsForFilter(EventDispatchThread.java:116)
	at java.awt.EventDispatchThread.pumpEventsForHierarchy(EventDispatchThread.java:105)
	at java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:101)
	at java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:93)
	at java.awt.EventDispatchThread.run(EventDispatchThread.java:82)
```

But if I let it to chose on its own, that is switch to CPU ( ---> sequentialComputation(g);  ), it works OK: it produces B&W image out of color one (BTW: there was no image.jpg included in the ZIP, I had to provide my own and also change the path value from "/tmp/" to "tmp/", else it would ends up in error saying it cannot find the image file!

SO NOW TO MY QUESTION: do I understand it right, that for the TornadoVM to work on GPU it needs its own native DLL to be included (just like other similar JAVA "products" that uses GPU, like JCuda)? If so, how to obtain them...where? They are not in the src master ZIP file, and Google was of no help at all too. Actually, under drivers, I saw some .cpp and .h files, but it is of no purpose for me as I have no knowledge how to use them and generate those system native DLLs (well, in case that is what they are).

Guys, it would be really helpfull a lot, if some of you would actually also provide COMPILED TornadoVM.jar file and TornadoVM-natives.jar (dlls) as well, cos it is really pain-in-da-eiss trying to make this work as it should...cos this has huge potential, like image renderers from 3D files - my case scenario (render that take several hours could be done in few minutes, right?)

**Comment by jjfumero** [comment]
You need to compile TornadoVM first. 

The TornadoVM build-script will create all JAR files as well as the dynamic libraries for OpenCL/PTX execution (`.so` on Linux/OSx) . Onced TornadoVM is built, there is a directory in `<TornadoVM-ROOT>/dist/tornado-sdk/tornado-sdk-0.11-dev-<ID>/lib/` with all dynamic libraries and `<TornadoVM-ROOT>/dist/tornado-sdk/tornado-sdk-0.11-dev-<ID>/share/java/` with all `jar` files.

Besides, we have an easy-installer script for TornadoVM to be used from the terminal. Assuming you have a supported OS and OpenCL/CUDA installed, this should be fine. 

https://github.com/beehive-lab/tornadovm-installer

We do not use Netbeans, so I can't tell the instructions to make it work. We use IntelliJ, and here we have the setup:
https://github.com/beehive-lab/TornadoVM/blob/master/assembly/src/docs/3_INTELLIJ.md 

All JAR files are available and can be found here:

https://github.com/beehive-lab/TornadoVM/tree/maven-tornadovm 

https://github.com/beehive-lab/TornadoVM/tree/maven-tornadovm/tornado

**Reply by fafayqa** [reply]
Thank you for the JARS, it should help a lot as I have it set up in NetBeans, all it was missing were those dynamic libs (*.dll) that I was unable compile from the source (CPP is not my cap-of-tea, ya know) - I will play with it today and post if it works.

**Reply by fafayqa** [reply]
So, unfortunatelly, there are no DLLs in any of the compiled JARs, which is the part I am missing still - I hoped they would be there as it is common with other such jars, like JCuda, Rootbeer etc.

Can you, please, provide those native dynamic libraries for Windows 7 x64 as well? Without those, I guess, I cannot use TornadoVM in NetBeans, unfortunately. And I have no clue how to compile those DLL myself (I see CPP files there, but I am not CPP guy, so...)

I would expect, probably, something like "tornadovm-natives-windows-x64.dll" (or at least that is how I understand it, that TornadoVM has something like this too, just like the other abovementioned JCuda and Rootbeer), right?

I guess it would be really nice if there was also some sort of pre-compiled dynamic libraries repository - just like those already compiled TornadoVM JARs you provided, where anyone can download them without doing the compile process himself, more so if he have no clue how to compile CPP files...just my few bits of thoughts for future).

**Reply by jjfumero** [reply]
We have instructions for Windows using [MSys2](https://www.msys2.org/). 

If you follow this process, you will get the dynamic libraries for Windows under the Linux development tools:

https://github.com/beehive-lab/TornadoVM/blob/master/assembly/src/docs/20_INSTALL_WINDOWS_WITH_GRAALVM.md 

As of August 2021, we do not have support for native Windows DLLs (but pull requests are welcome ;-) )

**Reply by fafayqa** [reply]
I have no clue what MSys2 is.

"If you follow this process, you will get the dynamic libraries for Windows under the Linux development tools:" - what does this means? Do I have to have Linux installed to be able producing Windows DLLs you said cannot be done (you do not have support for it, as you wrote)? I am confused from your answers, really.

Still, if they can be done I do not understand why don't you provide them already compiled as I said in my previous reply and you did not answer that (like why don't you provide ZIP with already compiled Windows native DLLs, if you can provide ZIP with already compiled JAR - it does not make any sense to me, cos as you may see by my interest, I would really very like to use them in NetBeans JAVA project(s)).

As for the Graal - I ave it as compiled JARs already in NetBeans, if that is what you mean.

So once again if I got it right:  there are no Windows native DLLs and they cannot be even compiled at this point in time (August 2021) as you do not have support for it yet - is this correct? Cos if it is correct (if I understood you right),there is no way for me to get the TornadoVM work under NetBeans.

Also I have no clue what the "pull requests" means...???

**Reply by stratika** [reply]
Hi @fafayqa,

Let's try to simplify the confusion that has been created. 

At first, we do not offer a binary distribution yet. So, TornadoVM has to be built by source, or by using the Docker images. However, in your case because you use Windows, we do not have any Docker images available.

1. TornadoVM on Windows can be built by using `MSYS2` and the steps are described [here](https://github.com/beehive-lab/TornadoVM/blob/master/assembly/src/docs/20_INSTALL_WINDOWS_WITH_GRAALVM.md) (as mentioned earlier by @jjfumero). This offers a native built that we have currently used with [IntelliJ IDEA](https://github.com/beehive-lab/TornadoVM/blob/master/assembly/src/docs/3_INTELLIJ.md). I believe that importing the TornadoVM code base in NetBeans as a maven project, should be similar.

I have two questions at this point, because it is not clear from your description how are you running the failing examples: 
Q1) Have you tried installing TornadoVM with `MSYS2` in your machine? In this case, the examples and unit-tests should work with the MSYS2 console. Do they fail?
Q2) Are you trying to run the examples from within NetBeans?

2. If you want to build TornadoVM natively in Windows OS, this is not supported and you will have to take the responsibility for it. Then, you are welcome to contribute this part to the TornadoVM code base. These steps are described [here](https://github.com/beehive-lab/TornadoVM/blob/master/CONTRIBUTING.md). In essence, you will have to provide a proposal that will describe the steps in a documented manner. This will be the initial step before actually merging the changes via a GitHub pull request.

**Reply by fafayqa** [reply]
_"At first, we do not offer a binary distribution yet. So, TornadoVM has to be built by source, or by using the Docker images. However, in your case because you use Windows, we do not have any Docker images available."_
OK, understood - thanks for proper explanation.

_"TornadoVM on Windows can be built by using MSYS2 and the steps are described here (as mentioned earlier by @jjfumero). This offers a native built that we have currently used with IntelliJ IDEA. I believe that importing the TornadoVM code base in NetBeans as a maven project, should be similar."_
Well, I have never heard of, nor tried MSYS2, so I have to first check it out and understood what it is and how to use it, unfortunately. :-(
Secondly: I have to this day headaches from any Maven project! To me they are mega overcomplicated as for the "manual/by-the-hand" creation and maintaining project in NetBeans, therefore I always use and create ANT project instead, where is everything easy and straightforward, like adding local JARS etc., in Maven projects one needs to be kind of professor to make this work, tho I understand it may be just a case of individual "test", if you like. Still, I do not understand Maven projects enough, I can barely load them into NetBeans and problems start when it starts download required jars from maven and it show up there are not there - to this day I did not find easy way how to add such missing jars manually by hand from my local depository (compared to ANT projects, where it is just right click and point to the jar, that's it - that is not the case with Maven). So any link to any maven-thing is beyond me....

_Q1) Have you tried installing TornadoVM with MSYS2 in your machine? In this case, the examples and unit-tests should work with the MSYS2 console. Do they fail?_
As I wrote above, I have no clue what that software is, so my answer is NO - all my JAVA work is done only in NetBeans IDE, and all project is ANT project (if I download some Maven project I have to manually create new ANT project and move/copy files manually to the right places as that is the only way I can manage the project with easy and with all of my "powers", LOL)

_Q2) Are you trying to run the examples from within NetBeans?_
Yes, that is exactly what I did/do.

_If you want to build TornadoVM natively in Windows OS, this is not supported and you will have to take the responsibility for it. Then, you are welcome to contribute this part to the TornadoVM code base. These steps are described here. In essence, you will have to provide a proposal that will describe the steps in a documented manner. This will be the initial step before actually merging the changes via a GitHub pull request._
:-)))))) Oh, no way - that is waaaay beyond my ability as a "JAVA-programmer".

Just to summarize what I do so we can come to some final conclusion if what I want is even possible to do at all, so that I do not bother you with things that can not be done (I want to make the code of my JAVA renderer being computed mostly on GPU instead of 6-core CPU):

1. I created new ANT project in NetBeans, let's say I called it "TornadoVM-0.10_1_JARs".
2. I manually added the source files from your original Maven project (*.pom)
3. I manually added all the needed JARs (well at least according to NetBeans, where it does not complain about any error of missing files or code), namely: graal-sdk-21.1.0.jar; ejm-ddense-0.41.jar; ejml-core-0.41.jar; lucene-core-8.9.0.jar; tornado-api-0.10.jar
4. **I WOULD** (if I had them, of course, which I don't as you know) manually extracted all the needed native dyn libs to folder called "dll" and added path to it via VM Options in project settings like this: ` -Djava.library.path="dll"`
5. I selected file called "EJMLMatrixMult.java" in the first examples folder and hit SHFT+F6, I got this error in NetBeans output window: 

```
Exception in thread "main" java.lang.NullPointerException
	at java.lang.Class.forName0(Native Method)
	at java.lang.Class.forName(Class.java:264)
	at uk.ac.manchester.tornado.api.runtime.TornadoAPIProvider.loadScheduleRuntime(TornadoAPIProvider.java:62)
	at uk.ac.manchester.tornado.api.TaskSchedule.<init>(TaskSchedule.java:86)
	at uk.ac.manchester.tornado.examples.EJMLMatrixMult.main(EJMLMatrixMult.java:74)
C:\Users\1\AppData\Local\NetBeans\Cache\12.4\executor-snippets\run.xml:111: The following error occurred while executing this line:
C:\Users\1\AppData\Local\NetBeans\Cache\12.4\executor-snippets\run.xml:94: Java returned: 1
BUILD FAILED (total time: 0 seconds)
```

**Comment by fafayqa** [comment]
So guys, I have followed the link you gave me about the COMPILATION of the TornadoVM under Windows using MySys2 and...voila: all seems to be done successfully and all tests form the MINGx64  I made passed! :-)

Thank you - being very happy I managed at least to compile it incl. those native dlls for Windows I was searching for so desperatelly - in case you would care to include them into your zip file here in GitHub for others like me who are struggling to obtain them, I can send those two dlls to you gladly, just tell me how.

Now, as I have those two, I will test it under my project in NetBeans finding out if that was really the case it did not work, missing those two native DLLs of TornadoVM - will give you update later today (fingers crossed once again that all will work in my NetBeans 12.4 x64)!

**UPDATE**

Unfortunately, despite now having those 2 TornadoVM native dyn libs ("tornado-opencl.dll" and "tornado-ptx.dll") included properly in the NetBenas project, I still get the error when trying to run test files from there:

```
Exception in thread "main" java.lang.NullPointerException
	at java.lang.Class.forName0(Native Method)
	at java.lang.Class.forName(Class.java:264)
	at uk.ac.manchester.tornado.api.runtime.TornadoAPIProvider.loadScheduleRuntime(TornadoAPIProvider.java:62)
	at uk.ac.manchester.tornado.api.TaskSchedule.<init>(TaskSchedule.java:86)
	at uk.ac.manchester.tornado.examples.EJMLMatrixMult.main(EJMLMatrixMult.java:74)
C:\Users\1\AppData\Local\NetBeans\Cache\12.4\executor-snippets\run.xml:111: The following error occurred while executing this line:
C:\Users\1\AppData\Local\NetBeans\Cache\12.4\executor-snippets\run.xml:94: Java returned: 1
BUILD FAILED (total time: 0 seconds)
```

So, to me it either means there are some other TornadoVM native dyn libs in need that were not compiled, OR something else like some other native dyn libs from some other JAR of different developer that are needed are not there (and I have feeling the latter might be the case, right? like maybe some DLLs from the Graal I downloaded and unpacked, hm?). If that is the true: could you tell me which other DLLs are needed to be included to my project so that the TornadoVM would work finally for me in NetBeans (as it works nicelly when run from the MINGx64 under MySys2)?

What I think is although I have correct graal JARs included in my project, isn't there also some native dyn libs of its own too that your TornadoVM expects? So that if I would know which one exactly those are, I would include them into my project copied directly from the installed Graal in my PC (the one that I downloaded when I was prepariong for the Tornado's dyn libs compilation process).

**Comment by stratika** [comment]
It is nice that the `MSYS2` way worked.

I hope the following parts will be useful:
1. ```tornado``` as a command used in a command prompt, it is an alias to the `java` binary that also contains various JDK parameters. So, when we want to use the IntelliJ IDE to run the examples (this is tested on a Linux environment), we have to configure the run configurations with the number of Java flags that the `tornado` command uses. This part is available [here](https://github.com/beehive-lab/TornadoVM/blob/master/assembly/src/docs/3_INTELLIJ.md#debugrun-configuration). But to be honest, it is not tested on Windows, so there may be more to be done.

All these flags can be displayed by running:
``` bash
tornado --printFlags
```
2. In my case, I have not tried to run TornadoVM examples through the IDE, which in my case is IntelliJ. So, I am not sure which libraries you will need to manually installed. This number may be high, depending on the JDK version that you use.

**Reply by fafayqa** [reply]
So, I added this to my NetBeans RUN VM Options:

`-server -XX:-UseJVMCIClassLoader -XX:-UseCompressedOops -Djava.ext.dirs=jar -Djava.library.path=dll -Dtornado.load.api.implementation=uk.ac.manchester.tornado.runtime.tasks.TornadoTaskSchedule -Dtornado.load.runtime.implementation=uk.ac.manchester.tornado.runtime.TornadoCoreRuntime -Dtornado.load.tornado.implementation=uk.ac.manchester.tornado.runtime.common.Tornado -Dtornado.load.device.implementation.opencl=uk.ac.manchester.tornado.drivers.opencl.runtime.OCLDeviceFactory -Dtornado.load.device.implementation.ptx=uk.ac.manchester.tornado.drivers.ptx.runtime.PTXDeviceFactory`

I also added all the JARs I found in the folder "D:\MyProjects\TornadoVM\bin\sdk\share\java\tornado\" created by the MySYS2 MINGx64 installation of TornadoVM + also set path to those 2 TornadoVM DLLs.

Now there is new error, this one:

```
Error: Could not create the Java Virtual Machine.
Error: A fatal exception has occurred. Program will exit.
Unrecognized VM option 'UseJVMCIClassLoader'
C:\Users\1\AppData\Local\NetBeans\Cache\12.4\executor-snippets\run.xml:111: The following error occurred while executing this line:
C:\Users\1\AppData\Local\NetBeans\Cache\12.4\executor-snippets\run.xml:94: Java returned: 1
BUILD FAILED (total time: 0 seconds)
```

I guess what it basically means is, that it cannot find that graal-jvmci stuff, as I think - mind you: I have never worked with these, so it is just my assumption here - that it is some sort of separate program, that needs to be installed/extracted somewhere on the HDD and then it has to be pointed to in these VM Options - am I right?

If so, can you tell me, how to tell in command line switches (flags) where the graal-jvmci is installed/extracted? I guess it expects this as there is this in the VM Options `-XX:-UseJVMCIClassLoader` which I think is telling the compiler to use graal-jvmci, but I have no jar of that name in my project, tho I have downloaded "graal-jvmci-8-master.zip" and extracted it to my HDD.

So it looks like all I have to add now is that graal-jvmci path to its folder to VM Options, but how to?

**Reply by jjfumero** [reply]
You need to run TornadoVM with a supported JDK (e.g., JDK 8 with JVMCI). Probably you need to configure Netbeans to use a different JDK. List of supported JDKs:

https://github.com/beehive-lab/TornadoVM/blob/master/assembly/src/docs/15_FAQ.md#list-of-compatible-jdks 

+ We recently added support for JDK 16.

**Comment by NicevilleWest** [comment]
Hello Juan, Thanks to you and your team for the wonderful work. I have a different native library question. I have an Ubuntu VM running in WSL2 on Microsoft Windows 10 Professional. I installed TornadoVM successfully using your automated scripts but the simple test of `tornado --devices` failed with the following error:

```bash
tornado --devices
#
# A fatal error has been detected by the Java Runtime Environment:
#
#  SIGSEGV (0xb) at pc=0x00007f0f2f6af230, pid=3436, tid=3437
#
# JRE version: OpenJDK Runtime Environment GraalVM CE 21.3.0 (11.0.13+7) (build 11.0.13+7-jvmci-21.3-b05)
# Java VM: OpenJDK 64-Bit Server VM GraalVM CE 21.3.0 (11.0.13+7-jvmci-21.3-b05, mixed mode, tiered, jvmci, jvmci compiler, parallel gc, linux-amd64)
# Problematic frame:
# C  0x00007f0f2f6af230
#
# No core dump will be written. Core dumps have been disabled. To enable core dumping, try "ulimit -c unlimited" before starting Java again
#
# An error report file with more information is saved as:
# /home/vero/gpu/TornadoVM-master/hs_err_pid3436.log
#
# If you would like to submit a bug report, please visit:
#   https://github.com/oracle/graal/issues

# The crash happened outside the Java Virtual Machine in native code.
# See problematic frame for where to report the bug.
#
/home/vero/gpu/TornadoVM-master/bin/bin/tornado: line 321:  3436 Aborted                 ${JAVA_CMD} ${JAVA_FLAGS} $@ uk.ac.manchester.tornado.drivers.TornadoDeviceQuery verbose
```

I am able to use the GPU with other applications without issues.
Any suggestions are very welcome.

**Reply by jjfumero** [reply]
Hi @NicevilleWest . 
   thank you for your interest in TornadoVM.  We haven't tried yet to run TornadoVM inside WSL2. For windows, we can run through 
[MSys2](https://www.msys2.org/). 

https://github.com/beehive-lab/TornadoVM/blob/master/assembly/src/docs/20_INSTALL_WINDOWS_WITH_GRAALVM.md 

Nevertheless, the error you are getting is related to GraalVM. 

Having said that, if you can already run GPU applications within WSL, (OpenCL or CUDA) maybe it should be possible to run TornadoVM as well. Could you please open an issue and provide more info, backend used ( OpenCL, PTX or SPIRV ), machine, JDK version, etc.? 

Issues: https://github.com/beehive-lab/TornadoVM/issues

**Reply by jjfumero** [reply]
It seems OpenCL is still not supported to run on discrete GPUs.
https://github.com/microsoft/WSL/issues/6951 

However, Intel recently supported OpenCL to run on Intel HD Graphics (integrated GPUs) 

https://devblogs.microsoft.com/commandline/oneapi-l0-openvino-and-opencl-coming-to-the-windows-subsystem-for-linux-for-intel-gpus/ 

As I said, we haven't tested this yet. If you want to run on Windows, we strongly recommend following the instructions [here](
https://github.com/beehive-lab/TornadoVM/blob/master/assembly/src/docs/20_INSTALL_WINDOWS_WITH_GRAALVM.md).

**Comment by NicevilleWest** [comment]
Thanks a million for that. I will take a look at that.

    On Monday, 7 March 2022, 09:27:23 CET, Juan Fumero ***@***.***> wrote:  
 
 


It seems OpenCL is still not supported to run on discrete GPUs.
microsoft/WSL#6951

However, Intel recently supported OpenCL to run on Intel HD Graphics (integrated GPUs)

https://devblogs.microsoft.com/commandline/oneapi-l0-openvino-and-opencl-coming-to-the-windows-subsystem-for-linux-for-intel-gpus/

As I said, we haven't tested this yet. If you want to run on Windows, we strongly recommend following the instructions here.

—
Reply to this email directly, view it on GitHub, or unsubscribe.
Triage notifications on the go with GitHub Mobile for iOS or Android.

**Comment by NicevilleWest** [comment]
I presumed there was an integrated Intel Graphics but there is none in the laptop. Graphics says Add-in and none displayed in Device Manager. Thanks for drawing my attention to the hardware which I haven´t had a chance to look at for a very long time.


    On Monday, 7 March 2022, 09:27:23 CET, Juan Fumero ***@***.***> wrote:  
 
 


It seems OpenCL is still not supported to run on discrete GPUs.
microsoft/WSL#6951

However, Intel recently supported OpenCL to run on Intel HD Graphics (integrated GPUs)

https://devblogs.microsoft.com/commandline/oneapi-l0-openvino-and-opencl-coming-to-the-windows-subsystem-for-linux-for-intel-gpus/

As I said, we haven't tested this yet. If you want to run on Windows, we strongly recommend following the instructions here.

—
Reply to this email directly, view it on GitHub, or unsubscribe.
Triage notifications on the go with GitHub Mobile for iOS or Android.
